keps/sig-api-machinery/2876-crd-validation-expression-language/README.md

```diff
@@ -0,0 +1,492 @@
+objects the webhook is configured to intercept.
+
+### Goals
+
+- Make CRDs more self-contained and declarative
+- Simplify CRD development
+- Simplify CRD operations
+
+### Non-Goals
+
+- Support for validation formats, immutability or unions. These are all valuable improvements
+  but can be addressed orthogonally in separate KEPs.
+- Eliminate the need for webhooks entirely. Webhooks will still be needed for
+  some use cases. For example, if a validation check requires making a network
+  request to some other system, it will still need to be implemented in a webhook.
+- Support all validation done on native Kubernetes types. For example, CRD structural schemas has
+  complex validation rules that we CEL cannot support due to the lask of support for arbitrarily
+  deeply nested terms (CEL cannot support recursive data types).
+- Change how Kubernetes built-in types are validated, defaulted and converted.
+
+## Proposal
+
+An inline expression language like [Common Expression Language (CEL)](https://github.com/google/cel-go) would be an
+excellent supplement to the current validation mechanism because it is sufficiently expressive to
+satisfy a large set of remaining uses cases that none of the above can solve.
+For example, cross-field validation use cases can only be solved using
+expressions or webhooks.
+
+`x-kubernetes-validator` extension will be added to CRD structural schemas to allow CEL validation expressions.
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1
+kind: CustomResourceDefinition
+...
+  schema:
+    openAPIV3Schema:
+      type: object
+      properties:
+        spec:
+          x-kubernetes-validator: 
+            - rule: "minReplicas <= maxReplicas"
+              message: "minReplicas cannot be larger than maxReplicas"
+          type: object
+          properties:
+            minReplicas:
+              type: integer
+            maxReplicas:
+              type: integer
+```
+
+- Each validator may have multiple validation rules.
+
+- Each validation rule has an optional 'message' field for the error message that
+will be surfaced when the validation rule evaluates to false.
+
+- The validator will be scoped to the location of the `x-kubernetes-validator`
+extension in the schema. In the above example, the validator is scoped to the
+'spec' field.
+
+- For OpenAPIv3 object types, the expression will have direct access to all the
+fields of the object the validator is scoped to.
+  
+- For OpenAPIv3 scalar types (integer, string & boolean), the expression will have
+access to the scalar data element the validator is scoped to.
+
+- For OpenAPIv3 list and map types, the expression will have access to the data
+element of the list or map.
+  
+- For immutability use case, validator will have access to the existing version of the object. 
+
+<<[UNRESOLVED @jpbetz @deads2k @lavalamp]>>
+When performing an update, should the live state of a field also be readable by expressions? This could be used,
```

Comment:
I think we should allow this.  Is there a known limitation that makes this impossible/impractical?  In the validation handlers themselves, we have the information.

Reply:
It's possible, and @cici37 is planning to replace this unresolved with a few comments in the KEP explaining the functionality.  Specific use cases this helps with totally welcome!

Reply:
This has been added to the KEP.